### PR TITLE
feat(rxAppNavItem): FRMW-804 Target & Href prefix for nav items

### PR DESCRIPTION
**Pull request by halkon**
JIRA: https://jira.rax.io/browse/FRMW-804

This PR allows for applications to define a set of injectables (constant, value, factory, service, provider) that will allow for further customization of the nav. 

* HREF prefixes for nav items
* target prefixes for nav items

In the future we could also add `target` as nav item definitions (not developed in this PR, just worth noting).

### LGTMs
- [x] Dev LGTM
- [x] Dev+Vidyo LGTM
- [x] QE LGTM

### CODE REVIEW

File: src/components/rxApp/scripts/rxAppNavItem.js [hunk 1]
```diff
@@ -57,7 +57,52 @@ angular.module('encore.ui.rxApp')
         scope: {
             item: '='
         },
-        controller: function ($scope, $location, rxVisibility, Permission) {
+        controller: function ($scope, $location, rxVisibility, Permission, $injector) {
+            /*
+             * @description Determines whether or not a nav item should have it's url prefixed
```
Comment (Droogans) [inline src/components/rxApp/scripts/rxAppNavItem.js]: ...should have *its* url prefixed...

File: src/components/rxApp/scripts/rxAppNavItem.js [hunk 2]
```diff
@@ -57,7 +57,52 @@ angular.module('encore.ui.rxApp')
...
+                // this avoids issues when we do have a prefix but really the nav item should not have
+                // any defined href, i.e. items that have subitems
+                if (_.isEmpty(url)) {
+                    return url;
+                }
+
+                // Check if we have a definition of NAV_ITEM_PREFIX, if so let's retrieve it and return the given URL
+                // appended to the prefix.  This allows applications like origin to prefix nav items, while not
+                // messing with nav items in the demo/documentation.
+                var prefix = $injector.has('NAV_ITEM_PREFIX') ? $injector.get('NAV_ITEM_PREFIX') : false;
```
Comment (Droogans) [inline src/components/rxApp/scripts/rxAppNavItem.js]: The docs don't say what it returns when there is no injector matching the name found. It just returns `* - the instance`.

Do you have to manually catch the case where there is no `'NAV_ITEM_PREFIX'` with a ternary expression?

File: src/components/rxApp/scripts/rxAppNavItem.spec.js [hunk 3]
```diff
@@ -1,6 +1,6 @@
 describe('encore.ui.rxApp', function () {
     describe('rxAppNavItem', function () {
-        var scope, compile, el, someProp, rxvisibility;
+        var scope, compile, el, someProp, rxvisibility, rxenvironmenturlfilter, provide;
```
Comment (Droogans) [inline src/components/rxApp/scripts/rxAppNavItem.spec.js]: Is there a reason these aren't camelCased?

**Comment by halkon** [comment]
* Add Documentation around this not being something that people should use
* Convert unit tests to use `_injectable_`
* Use `document.createElement('a')` to do better URL construction
*

**Comment by parlarjb** [comment]
LGTM

File: src/components/rxApp/scripts/rxAppNavItem.js [hunk 4]
```diff
@@ -57,7 +57,58 @@ angular.module('encore.ui.rxApp')
...
+
+                // Check if we have a definition of NAV_ITEM_PREFIX, if so let's retrieve it and return the given URL
+                // appended to the prefix.  This allows applications like origin to prefix nav items, while not
+                // messing with nav items in the demo/documentation.
+                //
+                // _This is *NOT* meant for general consumption, this is strictly for the Origin Project_
+                // _This will eventually be deprecated and removed_
+                //
+
+                var prefix = $injector.has('NAV_ITEM_PREFIX') ? $injector.get('NAV_ITEM_PREFIX') : false;
```
Comment (CITguy) [inline src/components/rxApp/scripts/rxAppNavItem.js]: creating a variable regardless of if it being used seems smelly, see suggestion on line 93

File: src/components/rxApp/scripts/rxAppNavItem.js [hunk 5]
```diff
@@ -57,7 +57,58 @@ angular.module('encore.ui.rxApp')
...
+                if (prefix) {
+                    prefix = urlUtils.parseUrl(prefix);
```
Comment (CITguy) [inline src/components/rxApp/scripts/rxAppNavItem.js]: overriding the type seems smelly, see suggestion on line 93

File: src/components/rxApp/scripts/rxAppNavItem.js [hunk 6]
```diff
@@ -57,7 +57,58 @@ angular.module('encore.ui.rxApp')
...
+                    return prefix.protocol.concat('//').concat(prefix.host).concat(url);
+                }
+
+                // Return as normal if no prefix
+                return url;
```
Comment (CITguy) [inline src/components/rxApp/scripts/rxAppNavItem.js]: The logic of 86-93 seems smelly. I'd rather see something like...

```javascript
if ($injector.has('NAV_ITEM_PREFIX')) {
  var prefix = urlUtils.parseUrl($injector.get('NAV_ITEM_PREFIX'));
  return prefix.protocol.concat('//').concat(prefix.host).concat(url);
} else {
  return url;
}
```

Comment (halkon) [inline src/components/rxApp/scripts/rxAppNavItem.js]: Do you really want the else? seems redundant.  Otherwise, 👍

Comment (halkon) [inline src/components/rxApp/scripts/rxAppNavItem.js]: by that question I mean:
```javascript
if ($injector.has('NAV_ITEM_PREFIX')) {
    var prefix = urlUtils.parseUrl($injector.get('NAV_ITEM_PREFIX'));
    return prefix.protocol.concat('//').concat(prefix.host).concat(url);
}

// Return as normal if no prefix
return url;
```

**Comment by CITguy** [comment]
LGTM

**Comment by CITguy** [comment]
@halkon, please squash

**Comment by halkon** [comment]
@CITguy squashed

**Pull request merged**